fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/postgresql/PostgreSqlCodeSystemDAO.java

```diff
@@ -21,84 +21,59 @@
-        // might create the entry after we selected and found nothing
-        Integer result = getCodeSystemId(codeSystem);
-
-        // Create the resource if we don't have it already (set by the continue handler)
-        if (result == null) {
-            try {
-                result = fhirRefSequenceDAO.nextValue();
-
-                String INS = "INSERT INTO code_systems (code_system_id, code_system_name) VALUES (?, ?)";
-                try (PreparedStatement stmt = getConnection().prepareStatement(INS)) {
-                    // bind parameters
-                    stmt.setInt(1, result);
-                    stmt.setString(2, codeSystem);
-                    stmt.executeUpdate();
-                }
-            }
-            catch (SQLException e) {
-                if ("23505".equals(e.getSQLState())) {
-                    // another thread snuck in and created the record, so we need to fetch the correct id
-                    result = getCodeSystemId(codeSystem);
-
-                    if (result == null) {
-                        // This would be truly weird, but we protect against it anyway
-                        throw new IllegalStateException("No code system returned after duplicate found!");
-                    }
-                }
-                else {
-                    throw new FHIRPersistenceDataAccessException("codeSystem=" + codeSystem, e);
-                }
-            }
-
-        }
-
-        // There's no way result can be null here, so we're OK returning an int
-        return result;
     }
 
     /**
-     * Read the id for the named type
-     * @param codeSystem
-     * @return the database id, or null if the named record is not found
+     * Calls a stored procedure to read the system contained in the passed Parameter in the Code_Systems table.
+     * If it's not in the DB, it will be stored and a unique id will be returned.
+     * @param systemName
+     *
+     * @return The generated id of the stored system.
      * @throws FHIRPersistenceDataAccessException
      */
-    protected Integer getCodeSystemId(String codeSystem) throws FHIRPersistenceDataAccessException {
-        Integer result;
-
-        String sql1 = "SELECT code_system_id FROM code_systems WHERE code_system_name = ?";
-
-        try (PreparedStatement stmt = getConnection().prepareStatement(sql1)) {
-            stmt.setString(1, codeSystem);
-            ResultSet rs = stmt.executeQuery();
-            if (rs.next()) {
-                result = rs.getInt(1);
-            }
-            else {
-                result = null;
+    @Override
+    public int readOrAddCodeSystem(String systemName) throws FHIRPersistenceDataAccessException   {
+        final String METHODNAME = "readOrAddCodeSystem";
+        log.entering(CLASSNAME, METHODNAME);
+
+        int systemId;
+        String currentSchema;
+        String stmtString;
+        String errMsg = "Failure storing code system id: name=" + systemName;
```

Comment:
I think logically this should go in the catch block.

Reply:
I agree